<a href="https://cognitiveclass.ai"><img src = "https://ibm.box.com/shared/static/9gegpsmnsoo25ikkbl4qzlvlyjbgxs5x.png" width = 400> </a>

<h1 align=center><font size = 5>Segmenting and Clustering Neighborhoods in Toronto</font></h1>

## Introduction

In this lab, i use the Notebook to build the code to scrape the following Wikipedia page, https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M, in order to obtain the data that is in the table of postal codes and to transform the data into a pandas dataframe. Also, i will use the Foursquare API to explore neighborhoods in Toronto. i will use the **explore** function to get the most common venue categories in each neighborhood, and then use this feature to group the neighborhoods into clusters. i will use the *k*-means clustering algorithm to complete this task. 

## Table of Contents

<div class="alert alert-block alert-info" style="margin-top: 20px">

<font size = 3>
    
1. <a href="#item1">Download Dependencies</a>

2. <a href="#item2">Scrape Dataset</a>

3. <a href="#item3">Explore Dataset</a>

4. <a href="#item4">Explore Neighborhoods in Toronto</a>
    
5. <a href="#item5">Analyze Each Neighborhood</a>

6. <a href="#item6">Cluster Neighborhoods</a>

7. <a href="#item7">Examine Clusters</a>    
</font>
</div>

## 1. Download Dependencies

In [42]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

#!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library

#install html import package
!conda install -c conda-forge lxml --yes

print('Libraries imported.')

Solving environment: done

# All requested packages already installed.

Solving environment: done

# All requested packages already installed.

Libraries imported.


## 2. Scrape Dataset

In [17]:
#read data from wikipedia page and display first 5 rows
df_toronto = pd.read_html('https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M')[0]
df_toronto.head()

Postal Code           Borough               Neighborhood
0         M1A      Not assigned               Not assigned
1         M2A      Not assigned               Not assigned
2         M3A        North York                  Parkwoods
3         M4A        North York           Victoria Village
4         M5A  Downtown Toronto  Regent Park, Harbourfront

In [18]:
# what is the shape of the data
print('The dataframe has {} rows and {} columns .'.format(
        df_toronto.shape[0],
       df_toronto.shape[1]
    )
)

The dataframe has 180 rows and 3 columns .


In [21]:
#Only process the cells that have an assigned borough. Ignore cells with a borough that is Not assigned
toronto_dropNotAssigned = df_toronto[df_toronto.Borough != 'Not assigned'].reset_index(drop=True)

#Handling existence of more than one neighborhood in one postal code area
toronto_grouped = toronto_dropNotAssigned.groupby(['Postal Code','Borough'], as_index=False).agg(lambda x: ','.join(x))

#If a cell has a borough but a Not assigned neighborhood, then the neighborhood will be the same as the borough
mask = toronto_grouped['Neighborhood'] == "Not assigned"
toronto_grouped.loc[mask, 'Neighborhood'] = toronto_grouped.loc[mask, 'Borough']

toronto_grouped.head(12)

Postal Code      Borough                                       Neighborhood
0          M1B  Scarborough                                     Malvern, Rouge
1          M1C  Scarborough             Rouge Hill, Port Union, Highland Creek
2          M1E  Scarborough                  Guildwood, Morningside, West Hill
3          M1G  Scarborough                                             Woburn
4          M1H  Scarborough                                          Cedarbrae
5          M1J  Scarborough                                Scarborough Village
6          M1K  Scarborough        Kennedy Park, Ionview, East Birchmount Park
7          M1L  Scarborough                    Golden Mile, Clairlea, Oakridge
8          M1M  Scarborough    Cliffside, Cliffcrest, Scarborough Village West
9          M1N  Scarborough                        Birch Cliff, Cliffside West
10         M1P  Scarborough  Dorset Park, Wexford Heights, Scarborough Town...
11         M1R  Scarborough                                  Wexford, Maryvale

In [22]:
# what is the shape of the data
print('The toronto grouped dataframe has {} rows and {} columns .'.format(
   toronto_grouped.shape[0],
    toronto_grouped.shape[1]
    )
)

The toronto grouped dataframe has 103 rows and 3 columns .


## 3. Explore Dataset

#### Use the Geocoder package or the csv file to add coodinates to toronto_grouped dataframe

In [25]:
#read Postal code coodinates from geospatial data in the provided link
geo_data = pd.read_csv('http://cocl.us/Geospatial_data')

#rename the Postal Code Column in the geodata dataframe to PostalCode
geo_data.rename(index=str, columns={"Postal Code": "PostalCode"}, inplace = True)
geo_data.head()

PostalCode   Latitude  Longitude
0        M1B  43.806686 -79.194353
1        M1C  43.784535 -79.160497
2        M1E  43.763573 -79.188711
3        M1G  43.770992 -79.216917
4        M1H  43.773136 -79.239476

In [26]:
#rename the Postal Code Column in toronto_grouped dataframe to PostalCode
toronto_grouped.rename(index=str, columns={"Postal Code": "PostalCode"}, inplace = True)
toronto_grouped.head()

PostalCode      Borough                            Neighborhood
0        M1B  Scarborough                          Malvern, Rouge
1        M1C  Scarborough  Rouge Hill, Port Union, Highland Creek
2        M1E  Scarborough       Guildwood, Morningside, West Hill
3        M1G  Scarborough                                  Woburn
4        M1H  Scarborough                               Cedarbrae

Now that build a dataframe of the postal code of each neighborhood along with the borough name and neighborhood name, in order to utilize the Foursquare location data, we need to get the latitude and the longitude coordinates of each neighborhood

In [27]:
#using table join function to combine 2 dataframes
toronto_neigh = pd.merge(toronto_grouped, geo_data, on='PostalCode', how='inner')
toronto_neigh.head()

PostalCode      Borough                            Neighborhood   Latitude  \
0        M1B  Scarborough                          Malvern, Rouge  43.806686   
1        M1C  Scarborough  Rouge Hill, Port Union, Highland Creek  43.784535   
2        M1E  Scarborough       Guildwood, Morningside, West Hill  43.763573   
3        M1G  Scarborough                                  Woburn  43.770992   
4        M1H  Scarborough                               Cedarbrae  43.773136   

   Longitude  
0 -79.194353  
1 -79.160497  
2 -79.188711  
3 -79.216917  
4 -79.239476

In [28]:

print('The dataframe has {} boroughs and {} neighborhoods.'.format(
        len(toronto_neigh['Borough'].unique()),
       toronto_neigh.shape[0]
    )
)

The dataframe has 10 boroughs and 103 neighborhoods.


#### Use geopy library to get the latitude and longitude values of Toronto.

In [29]:
address = 'Toronto, TO'

geolocator = Nominatim(user_agent="to_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto City are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto City are 43.6534817, -79.3839347.


#### Create a map of New York with neighborhoods superimposed on top.

In [30]:
# create map of New York using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip( toronto_neigh['Latitude'],  toronto_neigh['Longitude'],  toronto_neigh['Borough'],  toronto_neigh['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

## 4. Explore Neighborhoods in Toronto

#### Define Foursquare Credentials and Version

In [33]:
CLIENT_ID = 'WFBUP3WRUGXGR1NLHGC3JM41GSJB31HKHVE0KIYNEILSH4EC' # your Foursquare ID
CLIENT_SECRET = 'OR04LTTTO230R5PWT5YB5I4SBW2CZC4M3GBACWMYEVLHE2D0' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: WFBUP3WRUGXGR1NLHGC3JM41GSJB31HKHVE0KIYNEILSH4EC
CLIENT_SECRET:OR04LTTTO230R5PWT5YB5I4SBW2CZC4M3GBACWMYEVLHE2D0


In [34]:
#Scarborough_data.loc[0, 'Neighborhood']

In [35]:
#neighborhood_latitude = Scarborough_data.loc[0, 'Latitude'] # neighborhood latitude value
#neighborhood_longitude = Scarborough_data.loc[0, 'Longitude'] # neighborhood longitude value

#neighborhood_name = Scarborough_data.loc[0, 'Neighborhood'] # neighborhood name

#print('Latitude and longitude values of {} are {}, {}.'.format(neighborhood_name, 
                                                             #  neighborhood_latitude, 
                                                             #  neighborhood_longitude))

#### Let's create a function to explore all the neighborhoods in Toronto

In [36]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

#### Now, let's get the top 40 venues that are in all Toronto neighborhoods are within a radius of 500 meters.

In [38]:
LIMIT = 40
toronto_venues = getNearbyVenues(names=toronto_neigh['Neighborhood'],
                                   latitudes=toronto_neigh['Latitude'],
                                   longitudes=toronto_neigh['Longitude']
                                  )



Malvern, Rouge
Rouge Hill, Port Union, Highland Creek
Guildwood, Morningside, West Hill
Woburn
Cedarbrae
Scarborough Village
Kennedy Park, Ionview, East Birchmount Park
Golden Mile, Clairlea, Oakridge
Cliffside, Cliffcrest, Scarborough Village West
Birch Cliff, Cliffside West
Dorset Park, Wexford Heights, Scarborough Town Centre
Wexford, Maryvale
Agincourt
Clarks Corners, Tam O'Shanter, Sullivan
Milliken, Agincourt North, Steeles East, L'Amoreaux East
Steeles West, L'Amoreaux West
Upper Rouge
Hillcrest Village
Fairview, Henry Farm, Oriole
Bayview Village
York Mills, Silver Hills
Willowdale, Newtonbrook
Willowdale, Willowdale East
York Mills West
Willowdale, Willowdale West
Parkwoods
Don Mills
Don Mills
Bathurst Manor, Wilson Heights, Downsview North
Northwood Park, York University
Downsview
Downsview
Downsview
Downsview
Victoria Village
Parkview Hill, Woodbine Gardens
Woodbine Heights
The Beaches
Leaside
Thorncliffe Park
East Toronto, Broadview North (Old East York)
The Danforth West, 

#### Let's check the size of the resulting dataframe

In [40]:
#View data structure
print(toronto_venues.shape)
toronto_venues.head()

(1531, 7)


Neighborhood  Neighborhood Latitude  \
0                          Malvern, Rouge              43.806686   
1                          Malvern, Rouge              43.806686   
2  Rouge Hill, Port Union, Highland Creek              43.784535   
3       Guildwood, Morningside, West Hill              43.763573   
4       Guildwood, Morningside, West Hill              43.763573   

   Neighborhood Longitude                  Venue  Venue Latitude  \
0              -79.194353                Wendy’s       43.807448   
1              -79.194353  Interprovincial Group       43.805630   
2              -79.160497  Royal Canadian Legion       43.782533   
3              -79.188711         RBC Royal Bank       43.766790   
4              -79.188711      G & G Electronics       43.765309   

   Venue Longitude        Venue Category  
0       -79.199056  Fast Food Restaurant  
1       -79.200378            Print Shop  
2       -79.163085                   Bar  
3       -79.191151                  Bank  
4       -79.191537     Electronics Store

Let's check how many venues were returned for each neighborhood

In [41]:
toronto_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Agincourt                                                               4   
Alderwood, Long Branch                                                  8   
Bathurst Manor, Wilson Heights, Downsview North                        22   
Bayview Village                                                         4   
Bedford Park, Lawrence Manor East                                      23   
Berczy Park                                                            40   
Birch Cliff, Cliffside West                                             4   
Brockton, Parkdale Village, Exhibition Place                           22   
Business reply mail Processing Centre, South Ce...                     17   
CN Tower, King and Spadina, Railway Lands, Harb...                     17   
Caledonia-Fairbanks                                                     4   
Canada Post Gateway Processing Centre                                  12   
Cedarbrae                                                               8   
Central Bay Street                                                     40   
Christie                                                               17   
Church and Wellesley                                                   40   
Clarks Corners, Tam O'Shanter, Sullivan                                12   
Cliffside, Cliffcrest, Scarborough Village West                         2   
Commerce Court, Victoria Hotel                                         40   
Davisville                                                             32   
Davisville North                                                        8   
Del Ray, Mount Dennis, Keelsdale and Silverthorn                        4   
Don Mills                                                              26   
Dorset Park, Wexford Heights, Scarborough Town ...                      7   
Downsview                                                              15   
Dufferin, Dovercourt Village                                           18   
East Toronto, Broadview North (Old East York)                           5   
Eringate, Bloordale Gardens, Old Burnhamthorpe,...                      8   
Fairview, Henry Farm, Oriole                                           40   
First Canadian Place, Underground city                                 40   
Forest Hill North & West, Forest Hill Road Park                         4   
Garden District, Ryerson                                               40   
Glencairn                                                               5   
Golden Mile, Clairlea, Oakridge                                         9   
Guildwood, Morningside, West Hill                                       7   
Harbourfront East, Union Station, Toronto Islands                      40   
High Park, The Junction South                                          23   
Hillcrest Village                                                       4   
Humber Summit                                                           2   
Humberlea, Emery                                                        1   
Humewood-Cedarvale                                                      4   
India Bazaar, The Beaches West                                         22   
Kennedy Park, Ionview, East Birchmount Park                             6   
Kensington Market, Chinatown, Grange Park                              40   
Kingsview Village, St. Phillips, Martin Grove G...                      1   
Lawrence Manor, Lawrence Heights                                       13   
Lawrence Park                                                           3   
Leaside                                                                33   
Little Portugal, Trinity                                               40   
Malvern, Rouge                                                          2   
Milliken, Agincourt North, Steeles East, L'Amor.

#### Let's find out how many unique categories can be curated from all the returned venues

In [43]:
print('There are {} uniques categories.'.format(len(toronto_venues['Venue Category'].unique())))

There are 245 uniques categories.


## 5. Analyze Each Neighborhood

Once we have all the different venues (exactly 1531) I will group them by their Neighborhood to make it more readable and more understandable, and once it is grouped, I will apply One Hot Encoding to extract the dummie variables of all of those buildings. I will do that to help me after to know what is the most frequency thing in each neighborhood. For this, I will use the following code

In [44]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighborhood'] = toronto_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

toronto_onehot.head()

Yoga Studio  Accessories Store  Airport  Airport Food Court  \
0            0                  0        0                   0   
1            0                  0        0                   0   
2            0                  0        0                   0   
3            0                  0        0                   0   
4            0                  0        0                   0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0               0                0                 0                    0   
1               0                0                 0                    0   
2               0                0                 0                    0   
3               0                0                 0                    0   
4               0                0                 0                    0   

   Antique Shop  Aquarium  Art Gallery  Art Museum  Arts & Crafts Store  \
0             0         0            0           0                    0   
1             0         0            0           0                    0   
2             0         0            0           0                    0   
3             0         0            0           0                    0   
4             0         0            0           0                    0   

   Asian Restaurant  Athletics & Sports  Auto Garage  Auto Workshop  \
0                 0                   0            0              0   
1                 0                   0            0              0   
2                 0                   0            0              0   
3                 0                   0            0              0   
4                 0                   0            0              0   

   BBQ Joint  Baby Store  Bagel Shop  Bakery  Bank  Bar  Baseball Field  \
0          0           0           0       0     0    0               0   
1          0           0           0       0     0    0               0   
2          0           0           0       0     0    1               0   
3          0           0           0       0     1    0               0   
4          0           0           0       0     0    0               0   

   Baseball Stadium  Basketball Court  Basketball Stadium  Beach  Beer Bar  \
0                 0                 0                   0      0         0   
1                 0                 0                   0      0         0   
2                 0                 0                   0      0         0   
3                 0                 0                   0      0         0   
4                 0                 0                   0      0         0   

   Beer Store  Belgian Restaurant  Bike Shop  Bistro  Boat or Ferry  \
0           0                   0          0       0              0   
1           0                   0          0       0              0   
2           0                   0          0       0              0   
3           0                   0          0       0              0   
4           0                   0          0       0              0   

   Bookstore  Boutique  Brazilian Restaurant  Breakfast Spot  Brewery  \
0          0         0                     0               0        0   
1          0         0                     0               0        0   
2          0         0                     0               0        0   
3          0         0                     0               0        0   
4          0         0                     0               0        0   

   Bridal Shop  Bubble Tea Shop  Burger Joint  Burrito Place  Bus Line  \
0            0                0             0              0         0   
1            0                0             0              0         0   
2            0                0             0              0         0   
3            0                0             0              0         0   
4            0                0             0              0         0   

   Bus Station  Business Service  

And let's examine the new dataframe size.

In [45]:
toronto_onehot.shape


(1531, 245)

#### Next, let's group rows by neighborhood and by taking the mean of the frequency of occurrence of each category

In [46]:
toronto_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()
toronto_grouped

Neighborhood  Yoga Studio  \
0                                           Agincourt     0.000000   
1                              Alderwood, Long Branch     0.000000   
2     Bathurst Manor, Wilson Heights, Downsview North     0.000000   
3                                     Bayview Village     0.000000   
4                   Bedford Park, Lawrence Manor East     0.000000   
5                                         Berczy Park     0.000000   
6                         Birch Cliff, Cliffside West     0.000000   
7        Brockton, Parkdale Village, Exhibition Place     0.000000   
8   Business reply mail Processing Centre, South C...     0.058824   
9   CN Tower, King and Spadina, Railway Lands, Har...     0.000000   
10                                Caledonia-Fairbanks     0.000000   
11              Canada Post Gateway Processing Centre     0.000000   
12                                          Cedarbrae     0.000000   
13                                 Central Bay Street     0.025000   
14                                           Christie     0.000000   
15                               Church and Wellesley     0.025000   
16            Clarks Corners, Tam O'Shanter, Sullivan     0.000000   
17    Cliffside, Cliffcrest, Scarborough Village West     0.000000   
18                     Commerce Court, Victoria Hotel     0.000000   
19                                         Davisville     0.000000   
20                                   Davisville North     0.000000   
21   Del Ray, Mount Dennis, Keelsdale and Silverthorn     0.000000   
22                                          Don Mills     0.000000   
23  Dorset Park, Wexford Heights, Scarborough Town...     0.000000   
24                                          Downsview     0.000000   
25                       Dufferin, Dovercourt Village     0.000000   
26      East Toronto, Broadview North (Old East York)     0.000000   
27  Eringate, Bloordale Gardens, Old Burnhamthorpe...     0.000000   
28                       Fairview, Henry Farm, Oriole     0.000000   
29             First Canadian Place, Underground city     0.000000   
30    Forest Hill North & West, Forest Hill Road Park     0.000000   
31                           Garden District, Ryerson     0.000000   
32                                          Glencairn     0.000000   
33                    Golden Mile, Clairlea, Oakridge     0.000000   
34                  Guildwood, Morningside, West Hill     0.000000   
35  Harbourfront East, Union Station, Toronto Islands     0.000000   
36                      High Park, The Junction South     0.000000   
37                                  Hillcrest Village     0.000000   
38                                      Humber Summit     0.000000   
39                                   Humberlea, Emery     0.000000   
40                                 Humewood-Cedarvale     0.000000   
41                     India Bazaar, The Beaches West     0.000000   
42        Kennedy Park, Ionview, East Birchmount Park     0.000000   
43          Kensington Market, Chinatown, Grange Park     0.000000   
44  Kingsview Village, St. Phillips, Martin Grove ...     0.000000   
45                   Lawrence Manor, Lawrence Heights     0.000000   
46                                      Lawrence Park     0.000000   
47                                            Leaside     0.000000   
48                           Little Portugal, Trinity     0.025000   
49                                     Malvern, Rouge     0.000000   
50  Milliken, Agincourt North, Steeles East, L'Amo...     0.000000   
51  Mimico NW, The Queensway West, South of Bloor,...     0.000000   
52                        Moore Park, Summerhill East     0.000000   
53       New Toronto, Mimico South, Humber Bay Shores     0.000000   
54           North Park, Maple Leaf Park, Upwood Park     0.000000   
55                  North Toronto West, Lawrence Park     0.052632   
56                Northwest, West Humber - Clairvil

#### Let's confirm the new size

In [47]:
toronto_grouped.shape

(95, 245)

#### Let's print each neighborhood along with the top 5 most common venues

In [48]:
num_top_venues = 5

for hood in toronto_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = toronto_grouped[toronto_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Agincourt----
                        venue  freq
0                      Lounge  0.25
1   Latin American Restaurant  0.25
2              Clothing Store  0.25
3              Breakfast Spot  0.25
4  Modern European Restaurant  0.00


----Alderwood, Long Branch----
            venue  freq
0     Pizza Place  0.25
1  Sandwich Place  0.12
2    Dance Studio  0.12
3     Coffee Shop  0.12
4            Pool  0.12


----Bathurst Manor, Wilson Heights, Downsview North----
                 venue  freq
0          Coffee Shop  0.09
1                 Bank  0.09
2          Pizza Place  0.05
3  Fried Chicken Joint  0.05
4             Pharmacy  0.05


----Bayview Village----
                 venue  freq
0                 Café  0.25
1   Chinese Restaurant  0.25
2                 Bank  0.25
3  Japanese Restaurant  0.25
4  Monument / Landmark  0.00


----Bedford Park, Lawrence Manor East----
                venue  freq
0         Coffee Shop  0.09
1          Restaurant  0.09
2  Italian Restaurant  0.09
3

#### Let's put that into a *pandas* dataframe

First, let's write a function to sort the venues in descending order.

In [110]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

Now let's create the new dataframe and display the top 10 venues for each neighborhood.

In [111]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood      1st Most Common Venue  \
0                                        Agincourt  Latin American Restaurant   
1                           Alderwood, Long Branch                Pizza Place   
2  Bathurst Manor, Wilson Heights, Downsview North                       Bank   
3                                  Bayview Village         Chinese Restaurant   
4                Bedford Park, Lawrence Manor East             Sandwich Place   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                Lounge        Clothing Store        Breakfast Spot   
1           Coffee Shop        Sandwich Place                   Pub   
2           Coffee Shop        Ice Cream Shop           Supermarket   
3   Japanese Restaurant                  Café                  Bank   
4           Coffee Shop            Restaurant    Italian Restaurant   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0          Dance Studio             Drugstore            Donut Shop   
1                   Gym          Dance Studio                  Pool   
2      Sushi Restaurant                 Diner            Restaurant   
3      Department Store          Dessert Shop    Dim Sum Restaurant   
4              Pharmacy           Pizza Place                   Pub   

  8th Most Common Venue      9th Most Common Venue 10th Most Common Venue  
0      Doner Restaurant                    Dog Run    Distribution Center  
1          Cupcake Shop                Curling Ice          Deli / Bodega  
2         Deli / Bodega  Middle Eastern Restaurant            Bridal Shop  
3                 Diner              Women's Store           Dance Studio  
4               Butcher                       Café       Sushi Restaurant

## 6. Cluster Neighborhoods

Run *k*-means to cluster the neighborhood into 5 clusters.

In [113]:
# set number of clusters
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([1, 4, 1, 1, 1, 1, 1, 1, 1, 1], dtype=int32)

Let's create a new dataframe that includes the cluster as well as the top 10 venues for each neighborhood.

In [114]:
# add clustering labels
#toronto_merged = toronto_merged.drop('Cluster Labels', axis=1)
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_merged = toronto_neigh

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

#toronto_merged['Cluster Labels']= toronto_merged['Cluster Labels'].astype('int64')

toronto_merged.head() # check the last columns!

PostalCode      Borough                            Neighborhood   Latitude  \
0        M1B  Scarborough                          Malvern, Rouge  43.806686   
1        M1C  Scarborough  Rouge Hill, Port Union, Highland Creek  43.784535   
2        M1E  Scarborough       Guildwood, Morningside, West Hill  43.763573   
3        M1G  Scarborough                                  Woburn  43.770992   
4        M1H  Scarborough                               Cedarbrae  43.773136   

   Longitude  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0 -79.194353             1.0  Fast Food Restaurant            Print Shop   
1 -79.160497             3.0                   Bar         Women's Store   
2 -79.188711             1.0     Electronics Store    Mexican Restaurant   
3 -79.216917             4.0           Coffee Shop     Korean Restaurant   
4 -79.239476             1.0   Fried Chicken Joint                Bakery   

  3rd Most Common Venue        4th Most Common Venue 5th Most Common Venue  \
0         Women's Store                 Cupcake Shop            Donut Shop   
1          Dance Studio  Eastern European Restaurant             Drugstore   
2        Breakfast Spot                 Intersection                  Bank   
3         Women's Store                  Curling Ice             Drugstore   
4                  Bank           Athletics & Sports      Hakka Restaurant   

  6th Most Common Venue 7th Most Common Venue       8th Most Common Venue  \
0      Doner Restaurant               Dog Run         Distribution Center   
1            Donut Shop      Doner Restaurant                     Dog Run   
2        Medical Center   Rental Car Location  Construction & Landscaping   
3            Donut Shop      Doner Restaurant                     Dog Run   
4       Thai Restaurant           Gas Station        Caribbean Restaurant   

  9th Most Common Venue 10th Most Common Venue  
0        Discount Store                  Diner  
1   Distribution Center         Discount Store  
2            Comic Shop       Doner Restaurant  
3   Distribution Center         Discount Store  
4          Dessert Shop          Deli / Bodega

Finally, let's visualize the resulting clusters

In [115]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighborhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

TypeError: list indices must be integers or slices, not float

In [94]:
#toronto_merged['Cluster Labels'].dtypes

In [116]:
#toronto_merged['Cluster Labels'] = toronto_merged['Cluster Labels'].astype(int)
toronto_merged['Cluster Labels'].dtypes

dtype('float64')

## 5. Examine Clusters

Now, you can examine each cluster and determine the discriminating venue categories that distinguish each cluster. Based on the defining categories, i can then assign a name to each cluster as shown below 

#### Parks 

In [117]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 0, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
14       Scarborough             0.0                  Park   
20        North York             0.0                  Park   
23        North York             0.0                  Park   
25        North York             0.0                  Park   
40         East York             0.0                  Park   
44   Central Toronto             0.0                  Park   
48   Central Toronto             0.0                  Park   
50  Downtown Toronto             0.0                  Park   
74              York             0.0                  Park   
98              York             0.0                  Park   

         2nd Most Common Venue 3rd Most Common Venue  \
14                  Playground      Cuban Restaurant   
20               Women's Store          Cupcake Shop   
23  Construction & Landscaping     Convenience Store   
25           Food & Drink Shop         Women's Store   
40                 Coffee Shop          Intersection   
44                 Swim School              Bus Line   
48                  Restaurant            Donut Shop   
50                       Trail            Playground   
74                         Bar         Women's Store   
98           Convenience Store         Women's Store   

          4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
14                   Donut Shop      Doner Restaurant               Dog Run   
20                   Donut Shop      Doner Restaurant               Dog Run   
23                Women's Store           Curling Ice            Donut Shop   
25                 Cupcake Shop            Donut Shop      Doner Restaurant   
40            Convenience Store           Curling Ice            Donut Shop   
44  Eastern European Restaurant             Drugstore            Donut Shop   
48             Doner Restaurant               Dog Run   Distribution Center   
50                  Curling Ice          Dance Studio         Deli / Bodega   
74               General Travel          Cupcake Shop            Donut Shop   
98                 Cupcake Shop            Donut Shop      Doner Restaurant   

   7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
14   Distribution Center        Discount Store                 Diner   
20   Distribution Center        Discount Store                 Diner   
23      Doner Restaurant               Dog Run   Distribution Center   
25               Dog Run   Distribution Center        Discount Store   
40      Doner Restaurant               Dog Run   Distribution Center   
44      Doner Restaurant               Dog Run   Distribution Center   
48        Discount Store                 Diner    Dim Sum Restaurant   
50      Department Store          Dessert Shop    Dim Sum Restaurant   
74      Doner Restaurant               Dog Run   Distribution Center   
98               Dog Run   Distribution Center        Discount Store   

   10th Most Common Venue  
14     Dim Sum Restaurant  
20     Dim Sum Restaurant  
23         Discount Store  
25                  Diner  
40         Discount Store  
44         Discount Store  
48           Dessert Shop  
50                  Diner  
74         Discount Store  
98                  Diner

#### Shopping Malls

In [118]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 1, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels      1st Most Common Venue  \
0         Scarborough             1.0       Fast Food Restaurant   
2         Scarborough             1.0          Electronics Store   
4         Scarborough             1.0        Fried Chicken Joint   
5         Scarborough             1.0                 Playground   
7         Scarborough             1.0                     Bakery   
8         Scarborough             1.0        American Restaurant   
9         Scarborough             1.0               Skating Rink   
10        Scarborough             1.0          Indian Restaurant   
11        Scarborough             1.0                Auto Garage   
12        Scarborough             1.0  Latin American Restaurant   
13        Scarborough             1.0                Pizza Place   
17         North York             1.0                Golf Course   
18         North York             1.0                Coffee Shop   
19         North York             1.0         Chinese Restaurant   
22         North York             1.0           Ramen Restaurant   
26         North York             1.0                Coffee Shop   
27         North York             1.0                Coffee Shop   
28         North York             1.0                       Bank   
29         North York             1.0       Caribbean Restaurant   
30         North York             1.0              Grocery Store   
31         North York             1.0              Grocery Store   
32         North York             1.0              Grocery Store   
33         North York             1.0              Grocery Store   
35          East York             1.0                Pizza Place   
36          East York             1.0                 Beer Store   
37       East Toronto             1.0                      Trail   
38          East York             1.0                Coffee Shop   
39          East York             1.0          Indian Restaurant   
41       East Toronto             1.0           Greek Restaurant   
42       East Toronto             1.0                       Park   
43       East Toronto             1.0                       Café   
45    Central Toronto             1.0                       Park   
46    Central Toronto             1.0                Coffee Shop   
47    Central Toronto             1.0             Sandwich Place   
51   Downtown Toronto             1.0                Coffee Shop   
52   Downtown Toronto             1.0               Burger Joint   
53   Downtown Toronto             1.0                Coffee Shop   
54   Downtown Toronto             1.0                       Café   
55   Downtown Toronto             1.0                       Café   
56   Downtown Toronto             1.0                Coffee Shop   
58   Downtown Toronto             1.0                Coffee Shop   
59   Downtown Toronto             1.0                       Park   
60   Downtown Toronto             1.0                Coffee Shop   
61   Downtown Toronto             1.0                       Café   
62         North York             1.0             Sandwich Place   
63    Central Toronto             1.0             Ice Cream Shop   
64    Central Toronto             1.0           Sushi Restaurant   
66   Downtown Toronto             1.0                       Café   
67   Downtown Toronto             1.0                       Café   
68   Downtown Toronto             1.0            Airport Service   
69   Downtown Toronto             1.0                       Café   
70   Downtown Toronto             1.0                       Café   
71         North York             1.0             Clothing Store   
72         North York             1.0                Pizza Place   
73               York             1.0                       Park   
75   Downtown Toronto             1.0              Grocery Store   
76       West Toronto             1.0                     Bakery   
77       West Toronto             1.0           Asian Restaurant   
78

#### Sports and Recreation

In [119]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 2, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
91   Etobicoke             2.0      Business Service        Baseball Field   
97  North York             2.0        Baseball Field         Women's Store   

   3rd Most Common Venue        4th Most Common Venue 5th Most Common Venue  \
91         Women's Store                    Drugstore            Donut Shop   
97          Dance Studio  Eastern European Restaurant             Drugstore   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
91      Doner Restaurant               Dog Run   Distribution Center   
97            Donut Shop      Doner Restaurant               Dog Run   

   9th Most Common Venue 10th Most Common Venue  
91        Discount Store                  Diner  
97   Distribution Center         Discount Store

#### Bars

In [120]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 3, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
1  Scarborough             3.0                   Bar         Women's Store   

  3rd Most Common Venue        4th Most Common Venue 5th Most Common Venue  \
1          Dance Studio  Eastern European Restaurant             Drugstore   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
1            Donut Shop      Doner Restaurant               Dog Run   

  9th Most Common Venue 10th Most Common Venue  
1   Distribution Center         Discount Store

#### Coffee Shops

In [121]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 4, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels  1st Most Common Venue  \
3         Scarborough             4.0            Coffee Shop   
6         Scarborough             4.0     Chinese Restaurant   
15        Scarborough             4.0   Fast Food Restaurant   
24         North York             4.0           Home Service   
34         North York             4.0  Portuguese Restaurant   
49    Central Toronto             4.0            Coffee Shop   
57   Downtown Toronto             4.0            Coffee Shop   
65    Central Toronto             4.0         Sandwich Place   
80               York             4.0            Coffee Shop   
85   Downtown Toronto             4.0            Coffee Shop   
86        Mississauga             4.0            Coffee Shop   
89          Etobicoke             4.0            Pizza Place   
95          Etobicoke             4.0      Convenience Store   
99          Etobicoke             4.0            Pizza Place   
100         Etobicoke             4.0         Sandwich Place   

    2nd Most Common Venue        3rd Most Common Venue 4th Most Common Venue  \
3       Korean Restaurant                Women's Store           Curling Ice   
6       Convenience Store                  Coffee Shop      Department Store   
15     Chinese Restaurant                  Coffee Shop              Pharmacy   
24            Coffee Shop                Grocery Store              Pharmacy   
34            Coffee Shop                  Pizza Place          Hockey Arena   
49                    Pub                   Bagel Shop        Sandwich Place   
57     Italian Restaurant              Bubble Tea Shop                  Café   
65                   Café                  Coffee Shop                  Park   
80         Discount Store            Convenience Store        Sandwich Place   
85       Sushi Restaurant                  Yoga Studio     College Cafeteria   
86                  Hotel    Middle Eastern Restaurant   American Restaurant   
89            Coffee Shop               Sandwich Place                   Pub   
95                   Park                 Liquor Store           Coffee Shop   
99     Chinese Restaurant                 Intersection        Sandwich Place   
100         Women's Store  Eastern European Restaurant            Donut Shop   

         5th Most Common Venue       6th Most Common Venue  \
3                    Drugstore                  Donut Shop   
6                   Hobby Shop                 Bus Station   
15                        Bank                 Pizza Place   
24                 Pizza Place                        Bank   
34                Intersection               Women's Store   
49       Vietnamese Restaurant            Sushi Restaurant   
57                 Yoga Studio  Modern European Restaurant   
65   Middle Eastern Restaurant                Liquor Store   
80               Women's Store                 Curling Ice   
85                    Beer Bar               Smoothie Shop   
86    Mediterranean Restaurant         Fried Chicken Joint   
89                         Gym                Dance Studio   
95                    Pharmacy                 Pizza Place   
99              Discount Store   Middle Eastern Restaurant   
100           Doner Restaurant                     Dog Run   

    7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
3        Doner Restaurant               Dog Run   Distribution Center   
6           Deli / Bodega          Dessert Shop    Dim Sum Restaurant   
15         Sandwich Place        Discount Store         Grocery Store   
24                Butcher          Dessert Shop          Dance Studio   
34           Dance Studio         Deli / Bodega      Department Store   
49     Light Rail Station          Liquor Store   American Restaurant   
57         Sandwich Place          Burger Joint      Ramen Restaurant   
65             Donut Shop          Burger Joint     Indian Restaurant   
80             Donut Shop      Doner Restaurant           

## Conclusion 

We can try to answer the question: So, if you would like to go to a Shopping Mall, where should you go?

So we can see that:
Cluster 0 is more the Parks , 
Cluster 1 is more commercial area where there are plenty of restaurants, and places to spend money. 
Cluster 2 is for sports and recreational services 
Cluster 3 is bar areas for drinkers 
Cluster 4 is for coffee shops 

The model could be improved by reducing the distance of proximity. But at least we can see the areas in the Shopping Malls are the places to go to when looking to do some shopping 

Another question would be which neighbor has all amneties close to each other and in this case its: **Scarborough
